doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+   is specified.
+
+   .. todo PR I see ? is sometimes need in the ltac2_expr and sometimes not, but
+      I don't when which is required
+
+   NOT RIGHT YET: In the :token:`ltac2_match_pattern`\s, metavariables have the form :n:`?@ident`, whereas
+   in the :n:`@ltac2_expr`\s, the question mark is omitted.  Choose your metavariable
+   names with care to avoid name conflicts.  For example, if you use the metavariable `S`,
+   then the :token:`ltac2_expr` can't use `S` to refer to the constructor of `nat`
+   without qualifying the constructor as `Datatypes.S`.
+
+   .. todo below: is matching non-linear unification?  is it the same or different
+      from unification elsewhere in Coq?
+
+   Matching is non-linear: if a
+   metavariable occurs more than once, each occurrence must match the same
+   expression.  Matching is first-order except on variables of the form :n:`@?@ident`
+   that occur in the head position of an application. For these variables,
+   matching is second-order and returns a functional term.
+
+   .. todo the `@?ident` form is in dangling_pattern_extension_rule, not included in the doc yet
+      maybe belongs with "Applications"
+
+   `lazy_match!`
+      Causes the match to commit to the first matching branch
+      rather than trying a new match if :n:`@ltac2_expr` fails.
+      :ref:`Example<ltac2_match_vs_lazymatch_ex>`.
+
+   `match!`
+      If :n:`@ltac2_expr` fails, continue matching with the next branch.
+      Failures in subsequent tactics (after the `match!`) will not cause selection
+      of a new branch.  Examples :ref:`here<ltac2_match_vs_lazymatch_ex>` and
+      :ref:`here<ltac2_match_vs_multimatch_ex>`.
+
+   `multi_match!`
+      If :n:`@ltac2_expr` fails, continue matching with the next branch.
+      When a :n:`@ltac2_expr` succeeds for a branch, subsequent failures
+      (after the `multi_match!`) causing consumption of all the successes
+      of :n:`@ltac2_expr` trigger selection of a new matching branch.
+      :ref:`Example<ltac2_match_vs_multimatch_ex>`.
+
+   :n:`@cpattern`
+      The syntax of :token:`cpattern` is
+      the same as that of :token:`term`\s, but it can contain pattern matching
+      metavariables in the form :n:`?@ident`.  :g:`_` can be used to match
+      irrelevant terms.
+
+      .. todo Example is broken :ref:`Example<ltac2_match_with_holes_ex>`.
+
+      .. todo Didn't understand the following 2 paragraphs well enough to revise
+         see https://github.com/coq/coq/pull/12103#discussion_r436297754 for a
+         possible example
+
+      When a metavariable in the form :n:`?id` occurs under binders,
+      say :n:`x__1, …, x__n` and the expression matches, the
+      metavariable is instantiated by a term which can then be used in any
+      context which also binds the variables :n:`x__1, …, x__n` with
+      same types. This provides with a primitive form of matching under
+      context which does not require manipulating a functional term.
```

Comment:
I thought Ltac2 failed to handle open terms entirely?  cc @ppedrot

Reply:
This is a cut-and-paste from Ltac1.  Perhaps it should be removed here?

Reply:
Yes.  Or just say something like "Unlike Ltac1, Ltac2 :n:`?id` metavariables only match closed terms." and drop a link to the corresponding paragraph in the Ltac1 documentation?

Reply:
I added that wording.  Ltac1 has the identical paragraph, so not sure what to link to.  Furthermore, we don't even define open and closed terms.  (Open means there's a metavariable in it??)